From 4c5fc48118ff339b3bd089ea51bf01f15c91e8c7 Mon Sep 17 00:00:00 2001
From: Arun Ravindran <arun.ravindran@intel.com>
Date: Mon, 19 Nov 2012 16:01:12 +0200
Subject: [PORT FROM MAIN] Telephony: fix "set up idle mode" + "Display" text icon

BZ: 15564

According to the 3GPP 31.111 chap 6.5.4: "if the terminal
receives an icon, and either an empty or no alpha text
string is given by the UICC, than the terminal shall reject
the command with general result "Command data not understood
by terminal".

This patch fix the code to respect this point for the set up
idle mode text and Display Text proactive command.

Change-Id: I2f5512eaf997c27d80ad35bde539050f2c94624f
Orig-Change-Id: Ic646f848915470f62c098498780ca8eb89204a3b
Category: aosp improvement
Domain: Telephony-Telephony-FWK
Origin: internal
Upstream-Candidate: yes
Signed-off-by: Nizar Haouati <nizar.haouati@intel.com>
Signed-off-by: Guillaume Lucas <guillaumex.lucas@intel.com>
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
---
 .../telephony/cat/CommandParamsFactory.java        |   27 ++++++++++++++-----
 1 files changed, 20 insertions(+), 7 deletions(-)

diff --git a/src/java/com/android/internal/telephony/cat/CommandParamsFactory.java b/src/java/com/android/internal/telephony/cat/CommandParamsFactory.java
index 0a21c28..95f5c5a 100644
--- a/src/java/com/android/internal/telephony/cat/CommandParamsFactory.java
+++ b/src/java/com/android/internal/telephony/cat/CommandParamsFactory.java
@@ -321,6 +321,14 @@ class CommandParamsFactory extends Handler {
         textMsg.isHighPriority = (cmdDet.commandQualifier & 0x01) != 0;
         textMsg.userClear = (cmdDet.commandQualifier & 0x80) != 0;
 
+        // According to 3GPP 31.111 chap 6.5.4 (ETSI TS 102 223 clause 6.5.4):
+        // If the terminal receives an icon and either an empty or no alpha
+        // text string is given by UICC, than the terminal shall reject the
+        // command with general result "Command not understood by terminal".
+        if ((iconId != null) && (textMsg.text == null)) {
+            throw new ResultException(ResultCode.CMD_DATA_NOT_UNDERSTOOD);
+        }
+
         mCmdParams = new DisplayTextParams(cmdDet, textMsg);
 
         if (iconId != null) {
@@ -355,15 +363,20 @@ class CommandParamsFactory extends Handler {
         if (ctlv != null) {
             textMsg.text = ValueParser.retrieveTextString(ctlv);
         }
-        // load icons only when text exist.
-        if (textMsg.text != null) {
-            ctlv = searchForTag(ComprehensionTlvTag.ICON_ID, ctlvs);
-            if (ctlv != null) {
-                iconId = ValueParser.retrieveIconId(ctlv);
-                textMsg.iconSelfExplanatory = iconId.selfExplanatory;
-            }
+
+        ctlv = searchForTag(ComprehensionTlvTag.ICON_ID, ctlvs);
+        if (ctlv != null) {
+            iconId = ValueParser.retrieveIconId(ctlv);
+            textMsg.iconSelfExplanatory = iconId.selfExplanatory;
         }
 
+        // According to 3GPP 31.111 chap 6.5.4 (ETSI TS 102 223 clause 6.5.4):
+        // If the terminal receives an icon and either an empty or no alpha
+        // text string is given by UICC, than the terminal shall reject the
+        // command with general result "Command not understood by terminal".
+        if ((iconId != null) && (textMsg.text == null)) {
+            throw new ResultException(ResultCode.CMD_DATA_NOT_UNDERSTOOD);
+        }
         mCmdParams = new DisplayTextParams(cmdDet, textMsg);
 
         if (iconId != null) {
-- 
1.7.4.1

